3_harmonize/src/clean_wqp_data.R

```diff
@@ -3,29 +3,171 @@
+#' 
+flag_missing_results <- function(wqp_data, commenttext_missing){
+  
+  wqp_data_out <- wqp_data %>%
+    mutate(flag_missing_result = case_when(
+      is.na(ResultMeasureValue) & is.na(DetectionQuantitationLimitMeasure.MeasureValue) ~ TRUE,
+      grepl("not reported", ResultDetectionConditionText, ignore.case = TRUE) ~ TRUE,
+      grepl(paste(commenttext_missing, collapse = "|"), ResultCommentText, ignore.case = TRUE) ~ TRUE,
+      TRUE ~ FALSE
+    ))
+  
+  return(wqp_data_out)
+  
+}
+
+
+
+#' @title Flag duplicated records
+#' 
+#' @description 
+#' Function to flag duplicated rows based on a user-supplied definition
+#' of a duplicate record. 
+#' 
+#' @param wqp_data data frame containing the data downloaded from the WQP, 
+#' where each row represents a data record.
+#' @param duplicate_definition character string(s) indicating which columns are
+#' used to identify a duplicate record. Duplicate records are defined as those 
+#' that share the same value for each column within `duplicate_definition`.
+#'
+#' @returns 
+#' Returns a data frame containing data downloaded from the Water Quality Portal,
+#' where each row represents a data record. New columns appended to the original
+#' data frame include flags for duplicated records. 
+#' 
+flag_duplicates <- function(wqp_data, duplicate_definition){
+  
+  # Flag duplicate records using the `duplicate_definition`
+  wqp_data_out <- wqp_data %>%
+    group_by(across(all_of(duplicate_definition))) %>% 
+    mutate(n_duplicated = n(),
+           flag_duplicated_row = if_else(n_duplicated > 1, TRUE, FALSE)) %>% 
+    ungroup()
+  
+  return(wqp_data_out)
+  
+}
+
+
+#' @title Remove duplicated records
+#' 
+#' @description
+#' Function to append additional flags to sets of duplicate rows that are then 
+#' used to drop duplicates from the dataset. Currently, we randomly retain the 
+#' first record in a set of duplicated rows and drop all others.
+#' 
+#' @param wqp_data data frame containing the data downloaded from the WQP, 
+#' where each row represents a data record.
+#' @param duplicate_definition character string(s) indicating which columns are
+#' used to identify a duplicate record. Duplicate records are defined as those 
+#' that share the same value for each column within `duplicate_definition`.
+#' 
+#' @returns 
+#' Returns a data frame containing data downloaded from the Water Portal in which
+#' duplicated rows have been removed. 
+#' 
+remove_duplicates <- function(wqp_data, duplicate_definition){
+
+  wqp_data_out <- wqp_data %>%
+    group_by(across(all_of(duplicate_definition))) %>% 
+    # To help resolve duplicates, randomly select the first record
+    # from each duplicated set and flag all others for exclusion.
+    mutate(dup_number = seq(n_duplicated),
+           flag_duplicate_drop_random = if_else(n_duplicated > 1 & 
+                                                  dup_number != 1, 
+                                                TRUE, FALSE)) %>%
```

Comment:
Same comment here - I think you can drop the `if_else()`

Reply:
So long, `if_else()`!